### ISSUE TITLE
B018 does not trigger on dictionary subscript

### ISSUE THREAD

**Issue by jbr-noba** [astral-sh/ruff#23932]
### Summary

I would have expected useless subscripts to get flagged the same way as useless attribute access, but it currently isn't the case:
```py
x.attr  # B018
x["item"]  # no error
```

In case you're wondering, this would have prevented a real bug. In a codebase, I found some code that looks like:
```py
my_value = my_dictionary["some_long_key"]["another_long_key"]
+my_dictionary["yet_another_long_key"]
-my_dictionary["again_another_key"]
```
The intention was clearly to assign the result of the operations to `my_value`, but, due to missing parentheses, only the first term is assigned and the following 2 lines are useless expressions.

I do realize there could be some corner cases where `__getitem__` has a side-effect, though I believe the [caveats in the B018 documentation](https://docs.astral.sh/ruff/rules/useless-expression/#known-problems) apply here as well and an explicit ignore/assignment is warranted. The same could be said about `__pos__` and `__neg__`, though they don't prevent an expression from being flagged:
```py
+x  # B018
-x  # B018
```

### Version

ruff 0.15.6 (e4c7f3577 2026-03-12)

**Comment by ntBre** [comment]
Thanks for the report! It looks like subscripts were added to our `contains_effect` helper in #1966 for the sake of a different rule. I think it might make sense to move that logic into that rule rather than considering all subscripts to contain side effects, as you said.

Alternatively, we could just expand the special cases in `B018`, like the one for attributes from https://github.com/astral-sh/ruff/pull/3455:

https://github.com/astral-sh/ruff/blob/34d05e2d3ef804bfde242e2333ab38585e81acc1/crates/ruff_linter/src/rules/flake8_bugbear/rules/useless_expression.rs#L103

**Comment by dscorbett** [comment]
This is another case where it would help for `contains_effect` to return three possible values (`Yes`, `Maybe`, `No`). B018 would raise a diagnostic for `Maybe` and `No`.
